flink-tests/src/test/java/org/apache/flink/test/checkpointing/ChangelogPeriodicMaterializationRescaleITCase.java

```diff
@@ -0,0 +1,126 @@
+    }
+
+    private void testRescale(int firstParallelism, int secondParallelism) throws Exception {
+        File checkpointFile = TEMPORARY_FOLDER.newFolder();
+        JobID firstJobID = generateJobID();
+
+        StreamExecutionEnvironment env =
+                getEnv(delegatedStateBackend, checkpointFile, 50, 0, 20, 0);
+        env.getCheckpointConfig()
+                .setExternalizedCheckpointCleanup(
+                        CheckpointConfig.ExternalizedCheckpointCleanup.RETAIN_ON_CANCELLATION);
+        env.setParallelism(firstParallelism);
+
+        JobGraph firstJobGraph =
+                buildJobGraph(
+                        env,
+                        new ControlledSource() {
+                            @Override
+                            protected void beforeElement(SourceContext<Integer> ctx)
+                                    throws Exception {
+                                if (currentIndex == TOTAL_ELEMENTS / 4) {
+                                    waitWhile(
+                                            () ->
+                                                    completedCheckpointNum.get() <= 0
+                                                            || getAllMaterializationId(
+                                                                            checkpointFile,
+                                                                            firstJobID)
+                                                                    .isEmpty());
+                                } else if (currentIndex > TOTAL_ELEMENTS / 4) {
+                                    throwArtificialFailure();
+                                }
+                            }
+                        },
+                        firstJobID);
+
+        try {
+            cluster.getMiniCluster().submitJob(firstJobGraph).get();
+            cluster.getMiniCluster().requestJobResult(firstJobGraph.getJobID()).get();
+        } catch (Exception ex) {
+            Preconditions.checkState(
+                    ExceptionUtils.findThrowable(ex, ArtificialFailure.class).isPresent());
+        }
+
+        env.setParallelism(secondParallelism);
+        JobID secondJobId = generateJobID();
+        JobGraph jobGraph =
+                buildJobGraph(
+                        env,
+                        new ControlledSource() {
+                            @Override
+                            protected void beforeElement(SourceContext<Integer> ctx)
+                                    throws Exception {
+                                if (currentIndex == TOTAL_ELEMENTS / 2) {
+                                    waitWhile(
+                                            () ->
+                                                    getAllMaterializationId(
+                                                                    checkpointFile, secondJobId)
+                                                            .isEmpty());
```

Comment:
Does this include also old (pre-recovery) materializations?
If so, the test currently validates recovery, but not the materialization after recovery.

Reply:
I have added the check condition to make sure there are materialization after recovery.